### ISSUE TITLE
#NewOre #NewTraits #NewAlloys Dilithium, Duranium, etc inspired by Star Trek

### ISSUE THREAD

**Issue by Flabort** [Zkaface/TAIGA#85]
#NewOre
Dilithium is an ore from Star Trek, that come as a crystaline substance. (Source: https://en.wikipedia.org/wiki/List_of_Star_Trek_materials) I proposed it be an ore.

Dilithium is primarily used as a power source, and can't be used for making tools. It's molten form is a powerful energy source, though, and can be used to fuel the smeltery; you should get much more molten dilithium from melting the ore down than you use up melting a stack of it.

#NewAlloy
Duranium is an alloy used in space-ship construction and for hand-held tools. This should be an alloy using Dilithium. I'm thinking 3 mb Dilithium, 3 mb molten Iron, and 1 mb Meteorite; this makes 2 mb Duranium. For stats, I'm thinking something like high durability, about 1600 for the head, 500 for extra, and 100 x2 for the handle. The drawbacks should include a lower mining speed, only diamond level mining, and low attack. A fully Duranium tool will have 4400 durability.

I'm thinking the extra should have the same trait as Duranite, Analysing, and the handle and extra should both have Cheap. The head though should have a new trait, Deflection II, and the handle should have Deflection I.

#NewTraits
Deflection I, Deflection II, Deflection III, Deflection IV, etc.
The trait has multiple levels, which stack (so a fully Duranium tool should have Deflection III).
When the tool is being held, either in the main hand or offhand, it reduces the damage by a set amount, that is, half a heart per level. It does so at the cost of it's own durability, though, with 10 durability lost per half a heart, though this can be reduced by Reinforced.

#NewAlloy
Tritanium is an even stronger, heavier alloy also used in the Star Trek universe for star ships and hand tools. My personal feelings on the matter are 5 mb Duranium, 3 mb Tritonite, and 3 mb Duranite; yielding 1 mb Tritanium. It shouldn't have quite as much durability as Duranium; 1200 for the head, 300 for extra, and 50 x2.5 for the handle. It should have a Duranite mining level, but still be slowish with low attack. It's traits should be Deflection III and super heavy for the head, Deflection I and super heavy for the extra, and Deflection I and super heavy for the handle. A fully Tritanium tool will have 3875 durability with Deflection V, preventing 2.5 hearts of damage per hit and using 50 durability per hit.

**Comment by Flabort** [comment]
My durability calculations assumed a basic 3-part tool. They did not count advanced 4-part tools or mattocks.

Assuming you use the stats I suggested, the the following applies for the advanced tools:
Duranium Mattock:  4800 durability, Deflection V
Tritanium Mattock:  4500 durability, Deflection VII
Duranium Hammer: 28800 durability, Deflection VII
Tritanium Hammer: 27000 durability, Deflection X
Duranium Scythe: 19200 durability, Deflection V
Tritanium Scythe: 18000 durability, Deflection VI
Duranium cleaver/battle axe: 18000 durability, Deflection VI
Tritanium cleaver/battle axe: 16875 durability, Deflection VIII
Duranium lumber axe/Excavator: 19200 durability, Deflection VI
Tritanium lumber axe/Excavator: 22500 durability, Deflection VIII

**Comment by Zkaface** [comment]
@Flabort Woah that's a lot of stuff :-D just woke up with a little hangover. 


We have a little problem here. If I put in another ore. It will be hard to find it again. An idea could be another dimension. But that would be a lot more work, because right now I do not understand how to create that. 

Durability is calculated automatically :-) there are several formulas for each tool. 

For the Trait: I am not sure if we are able to have a Trait which is able to stack if it's a full tool. I have to look for it. Maybe it's possible to have a workaround for this or I use an alternative way for tool level growing. 

Durability

**Comment by Flabort** [comment]
Yeah, I calculated those estimated durability using those formulas; wasn't sure which parts counted as heads and which ones as extras, though, which is why I say these are estimates.

So as for levels for the Trait, um... I think [Writable](https://github.com/SlimeKnights/TinkersConstruct/blob/master/src/main/java/slimeknights/tconstruct/tools/traits/TraitWritable.java) has multiple levels that stack. It extends [AbstractTraitLeveled](https://github.com/SlimeKnights/TinkersConstruct/blob/master/src/main/java/slimeknights/tconstruct/library/traits/AbstractTraitLeveled.java) which ALSO implies traits can have levels like modifiers. 

The "int maxlevels, int levels" bit tells me that you CAN have different parts have different levels that they apply, and you can set a maximum level for the trait, too. 

Paper heads USED to give 2 levels of writable, based on this [update to writable](https://github.com/SlimeKnights/TinkersConstruct/commit/5d03035b96507fa89849624b2e672800e17b8c2a) but they didn't update the materials to match so [the materials](https://github.com/SlimeKnights/TinkersConstruct/blob/master/src/main/java/slimeknights/tconstruct/tools/TinkerMaterials.java) still show how to set multiple levels in the code:

`    paper.setCraftable(true);
    paper.addItem("paper", 1, Material.VALUE_Fragment);
    paper.setRepresentativeItem(Items.PAPER);
    paper.addTrait(writable2, HEAD);
    paper.addTrait(writable);`

So you don't have to use new code, just use AbstractTraitLeveled instead of AbstractTrait. :-)

**Comment by Flabort** [comment]
I am just guessing, but it might look something like this.


> public class TraitDeflecting extends AbstractTraitLeveled {
> 
>     public TraitDeflecting(int levels) {
>         super("Deflecting", String.valueOf(levels), 0xaaaaaa, 6, levels)
>     }
>     
>     /* I have no idea, this is where stuff gets real */
>     /* I think you would override or subscribe to onArmorTick for when this needs to do stuff */
>     /* And simply multiply the effect by this.levels as far as I can tell */
> }

**Comment by Zkaface** [comment]
But this is not handled by each tool-part of a tool... at least it won't seem so. I gonna check that.

**Comment by Flabort** [comment]
> But this is not handled by each tool-part of a tool... at least it won't seem so. I gonna check that.

I might be completely wrong; I'm only learning still, after all. However, I think there might be other examples to support my hypothesis, in the form of [Magnetic II](https://github.com/SlimeKnights/TinkersConstruct/blob/master/src/main/java/slimeknights/tconstruct/tools/traits/TraitMagnetic.java) and [Crude II](https://github.com/SlimeKnights/TinkersConstruct/blob/master/src/main/java/slimeknights/tconstruct/tools/traits/TraitCrude.java). I would like to point out that all three leveled traits have two copies registered [here in the master traits list](https://github.com/SlimeKnights/TinkersConstruct/blob/master/src/main/java/slimeknights/tconstruct/tools/TinkerTraits.java).

However, it seems you are more correct than I. Looking a Flint tools in the Materials and You book, they should have a level of Crude IV (2+1+1), but instead have Crude III (2+1). Parts with the same level don't stack with each other, but will stack with different levels. Same with Iron, it only gets to Magnetic III. This is why Paper tools only get two extra modifiers despite having a max level of 3, only Writable1 and Writable2 are registered (though both only give 1 level). 

This is... actually fine. A fully duranium tool would only get Deflecting 3 no matter what kind it was, a fully Tritanium tool would only get 4, and you'd need to have a tool with 2 heads with a mix of Duranium and Tritanium to get any higher (which would get you level 6). This is actually a good limiting factor, and I don't mind it at all.

**Comment by Zkaface** [comment]
@Flabort You were right.... But! A tool part can hold up one level, but only one at a time and each level must be given via a separate variable. 

There are 3 regular tool types and 3 for bow and stuff. Let's say we offer 3 levels,from one to three (for head, handle and extra) then a full tool has Level 6. 3 + 2 + 1.

A hammer just has level 5 because the hammer head and the plate offer level 3 and the thick handle offers 2+1, but is only able to use the first one. 

You see its a bit complicated and I have to look what is possible and what is not.

**Comment by IdleGandalf** [comment]
Lets try and keep this names simple, no need for ridiculous titles.

**Comment by Flabort** [comment]
@IdleGandalf 
> Lets try and keep that names simple, no need for ridiculous titles, we are not on a social network. Short and on point.

I was using hashtags due to having read the following quote in the thread for Issue #67
quote by @Zkaface 
> Please open a separate Issue each Trait :) take a hastag to the thread name: #newtrait: name

**Comment by Zkaface** [comment]
@Flabort This was my fault. True! I had a discussion with @IdleGandalf about that. I lost ;-)
